independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java

```diff
@@ -199,7 +202,29 @@ static Type getProviderType(ClassInfo classInfo) {
         return restrictBeanTypes(types, beanDeployment.getAnnotations(classInfo));
     }
 
-    static Set<Type> getTypeClosure(ClassInfo classInfo, Map<TypeVariable, Type> resolvedTypeParameters,
+    static Set<Type> getTypeClosure(ClassInfo classInfo,
+            Map<TypeVariable, Type> resolvedTypeParameters,
+            BeanDeployment beanDeployment, BiConsumer<ClassInfo, Map<TypeVariable, Type>> resolvedTypeVariablesConsumer) {
+        return getTypeClosure(classInfo, null, null, resolvedTypeParameters, beanDeployment,
+                resolvedTypeVariablesConsumer);
+    }
+
+    static Set<Type> getTypeClosure(ClassInfo classInfo, FieldInfo producerField,
+            Map<TypeVariable, Type> resolvedTypeParameters,
+            BeanDeployment beanDeployment, BiConsumer<ClassInfo, Map<TypeVariable, Type>> resolvedTypeVariablesConsumer) {
+        return getTypeClosure(classInfo, null, producerField, resolvedTypeParameters, beanDeployment,
+                resolvedTypeVariablesConsumer);
+    }
+
+    static Set<Type> getTypeClosure(ClassInfo classInfo, MethodInfo producerMethod,
+            Map<TypeVariable, Type> resolvedTypeParameters,
+            BeanDeployment beanDeployment, BiConsumer<ClassInfo, Map<TypeVariable, Type>> resolvedTypeVariablesConsumer) {
+        return getTypeClosure(classInfo, producerMethod, null, resolvedTypeParameters, beanDeployment,
+                resolvedTypeVariablesConsumer);
+    }
+
+    static Set<Type> getTypeClosure(ClassInfo classInfo, MethodInfo producerMethod, FieldInfo producerField,
```

Comment:
I would replace the `MethodInfo/FieldInfo` with `AnnotationTarget` so that we wouldn't have to introduce two more getTypeClose() signatures...